ui-v2/vendor/metrics-providers/prometheus.js

```diff
@@ -276,59 +279,102 @@
       return all
     },
 
-    reformatSeries: function(response) {
-      // Handle empty results from prometheus.
-      if (!response || !response.data || !response.data.result
-        || response.data.result.length < 1) {
-        return [];
-      }
-      // Reformat the prometheus data to be the format we want which is
-      // essentially the same but with Date objects instead of unix timestamps.
-      return response.data.result[0].values.map(function(val){
-        return [new Date(val[0]*1000), parseFloat(val[1])]
-      })
-    },
-
-    fetchRequestRateSeries: function(serviceName, options){
-      var q = `sum(irate(envoy_listener_http_downstream_rq_xx{local_cluster="${serviceName}",envoy_http_conn_manager_prefix="public_listener_http"}[10m]))`
-      return this.fetchSeries(q, options).then(this.reformatSeries, function(xhr){
-        // Failure. log to console and return an blank result for now.
-        console.log("ERROR: failed to fetch requestRate", xhr.responseText)
-        return []
-      })
-    },
+    reformatSeries: function(unitSuffix, labelMap) {
+      return function(response) {
+        // Handle empty result sets gracefully.
+        if (!response.data || !response.data.result || response.data.result.length == 0
+            || !response.data.result[0].values
+            || response.data.result[0].values.length == 0) {
+          return emptySeries;
+        }
+        // Reformat the prometheus data to be the format we want with stacked
+        // values as object properties.
 
-    fetchErrorRateSeries: function(serviceName, options){
-      // 100 * to get a result in percent
-      var q = `sum(`+
-          `irate(envoy_listener_http_downstream_rq_xx{`+
-            `local_cluster="${serviceName}",`+
-            `envoy_http_conn_manager_prefix="public_listener_http",`+
-            `envoy_response_code_class="5"}[10m]`+
-          `)`+
-        `)`;
-      return this.fetchSeries(q, options).then(this.reformatSeries, function(xhr){
-        // Failure. log to console and return an blank result for now.
-        console.log("ERROR: failed to fetch errorRate", xhr.responseText)
-        return []
-      })
+        // Populate time values first based on first result since Prometheus will
+        // always return all the same points for all series in the query.
+        let series = response.data.result[0].values.map(function(d, i) {
+          return {
+            time: new Date(d[0]*1000),
```

Comment:
I meant to ask - I wonder if its worth keeping the provider to basic values (so numbers/strings/objects/arrays) - basically JSON, and converting to a date object in the UI component itself. That way a provider feels very much like an API

Reply:
Yeah I wasn't sure what was better here so open to changing that. It seemed like since it was a JS interface and not an HTTP API Date object would be more canonical.

If we pick a string, what format should we choose? Prometheus defaults to unix timestamps (seconds since epoch but as a float including milliseconds as a fraction). JS default is millisecond unix time. JSON default is ISO 8601 string. Which do you think is the most natural?

Reply:
Sounds like millisecond unix time is what is recorded so I think that maybe?